### PR TITLE
Add The Witness

### PR DESCRIPTION
**Pull request by NewSoupVi**
Adds The Witness Randomizer to Archipelago.

### CODE REVIEW

File: worlds/witness/__init__.py
Comment (black-sliver) [inline worlds/witness/__init__.py:109]: if this is already planned, i'd add an assert so you don't forget to remove this once the option exists. alternatively you could move it above the for loop, then it will be overwritten once the option is avalable

File: worlds/witness/__init__.py
Comment (black-sliver) [inline worlds/witness/__init__.py:121]:
```suggestion
    def get_filler_item_name(self) -> str:  # Used ny itemlinks
```

File: worlds/witness/__init__.py
Comment (black-sliver) [inline worlds/witness/__init__.py:124]: can this `random.choices()[0]` (plural) be replaced by a `random.choice()` (singular)?

File: worlds/witness/__init__.py
Comment (black-sliver) [inline worlds/witness/__init__.py:134]: we do 120 chars per line in AP. no need to add a line break here
```suggestion
    def __init__(self, player: int, name: str, address: typing.Optional[int], parent, ch_hex: int = -1):
```

File: worlds/witness/items.py
Comment (black-sliver) [inline worlds/witness/items.py:80]: two things here:
1. self.GOOD_ITEMS should probably be defined (maybe not assigned, but defined) earlier
2. the use of indentation here is hard to read. i'd have the if on a single line, not break after the `(` (120 chars per line is the limit)

File: worlds/witness/locations.py
Comment (black-sliver) [inline worlds/witness/locations.py:11]: purely stylistic, since all other docstrings in here look like this
```suggestion
    """
    Class that handles Witness Locations
    """
```

File: worlds/witness/regions.py
Comment (black-sliver) [inline worlds/witness/regions.py:30]:
```suggestion
    def connect(self, world: MultiWorld, player: int, source: str, target: str, panel_hex_to_solve_set=None):
        """
```

File: worlds/witness/rules.py
Comment (black-sliver) [inline worlds/witness/rules.py:32]: kewk

File: worlds/witness/rules.py
Comment (black-sliver) [inline worlds/witness/rules.py:179]: can you have a slot without completion_condition? should probably be added to to api.md then

File: worlds/witness/rules.py
Comment (black-sliver) [inline worlds/witness/rules.py:159]:
```suggestion
def set_rules(world: MultiWorld, player: int, logic: WitnessLogic, locat: WitnessLocations):
```

**Review by black-sliver — CHANGES_REQUESTED** [review]
Read it all, have not tested anything. A lot of the suggestions are just stylistic, probably missed some, so maybe also ask pycharm.

File: worlds/witness/rules.py
Comment (NewSoupVi) [reply worlds/witness/rules.py:178]: I don't remember where this came from, but it's a case of "It was there before". Idk from whom, or why. I can remove it's not necessary

File: worlds/witness/__init__.py
Comment (NewSoupVi) [reply worlds/witness/__init__.py:131]: Removed unnecessary line break

File: WebHostLib/static/assets/tutorial/The Witness/setup_en.md
Comment (NewSoupVi) [reply WebHostLib/static/assets/tutorial/The Witness/setup_en.md:17]: KEKW

File: worlds/witness/rules.py
Comment (black-sliver) [reply worlds/witness/rules.py:178]: @Berserker66 think you are most qualified to answer my question

File: worlds/witness/rules.py
Comment (black-sliver) [inline worlds/witness/rules.py:137]: as per https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/api.md#logic-mixin all methods you add to state through LogicMixin should be prefixed by `_witness_`. You can use pycharm's rightclick -> refactor -> rename to do that.

Basically whatever you add in a class that inherits from LogicMixin is added to the "global" State class. To avoid naming collisions with other worlds, we add a world-specific prefix.

**Review by black-sliver — APPROVED** [review]
Things I found were addressed, another set of eyes should probably still check the code.
Once this is final, please generate a couple of solo and multiworlds with this to make sure it does not crash and burn :-)

**Comment by Berserker66** [comment]
in init line 29 you create a new ParsedWitnessLogic object, and then in generate_early you create another one for each player.
The init of that class reads in and parses text files. This ends up being a repeated operation, causing many file fetches and not all that efficient parsing. This should be limited to running once, either during AP startup or on first world creation. Given that you need IDs from these files it'll probably end up being startup.

**Comment by NewSoupVi** [comment]
> in init line 29 you create a new ParsedWitnessLogic object, and then in generate_early you create another one for each player. The init of that class reads in and parses text files. This ends up being a repeated operation, causing many file fetches and not all that efficient parsing. This should be limited to running once, either during AP startup or on first world creation. Given that you need IDs from these files it'll probably end up being startup.

That makes sense. I guess I was lazy and trying to avoid bugs cause by forgetting to redo something that needs to be redone, operating with stale data from a world with different option. But you're right, I should bite the bullet and do it.

**Comment by NewSoupVi** [comment]
Ok, I've now seperated out every "data class" (logic, items, locations) I have into a static class for elements that don't change and a player-specific class for elements that change based on options. This should address all the issues. I think it's *slightly* improved generation time for Multiworlds as well, which feels rewarding, at least.

**Review by Berserker66 — APPROVED** [review]

**Comment by Berserker66** [comment]
So, this PR has over 100 commits and 3 authors. Normally, when there's multiple commits we like to squash, but with 3 authors the individual authorship info gets lost. So, we're kind of in a position of either "spam" the history with 100 commits or lose correct author info for the folder. Not sure how to proceed.

**Comment by NewSoupVi** [comment]
In what way would it get lost? Everyone would get credited for the whole thing, or just me for the whole thing?

**Comment by Berserker66** [comment]
https://github.blog/changelog/2019-12-19-improved-attribution-when-squashing-commits/#:~:text=Before%20today,%20whoever%20opened%20the,author%20on%20the%20squash%20commit.

**Comment by black-sliver** [comment]
My per-line co-authorship can be dropped since they did not really change any code, just style.
So we should probably ping @JarnoWesthof to give the OK if we want to squash it.

Rewriting history / partial squashing before merge would only keep the network graph alive i think, because everything else was touched during cleanup.

**Comment by JarnoWesthof** [comment]
Fine by me to squish it, i already expected that and i already got plenty of commits in the AP repo

**Pull request merged**
